### ISSUE TITLE
[Feature] `reasoning_content` in API for reasoning models like DeepSeek R1

### ISSUE THREAD

**Issue by gaocegege** [vllm-project/vllm#12468]
### 🚀 The feature, motivation and pitch

To better support reasoning models like DeepSeek-R1, adding a [`reasoning_content`](https://api-docs.deepseek.com/guides/reasoning_model) parameter to the API would be great so that users can see the steps in the reasoning process.

Ref https://github.com/sgl-project/sglang/issues/3043

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by gaocegege** [comment]
For DeepSeek series models, this means we could shift the `<think></think>` tags into the `reasoning_content`.

**Comment by simon-mo** [comment]
So this does break OpenAI compatibility but I think it is the right time to break it. Do you have suggestion on how to automatically figure out the `<think>` token so it's a bit general for the future model?

**Comment by gaocegege** [comment]
The OpenAI Python library still runs fine, even if we’re kinda messing with the compatibility a bit.

I think we could keep it more general since the <think> token might change for different models. I'm diving into the OpenAI server wrappers to raise a basic design proposal here.

```diff
from openai import OpenAI
client = OpenAI(api_key="", base_url="https://api.deepseek.com")

messages = [{"role": "user", "content": "9.11 and 9.8, which is greater?"}]
response = client.chat.completions.create(
    model="deepseek-reasoner",
    messages=messages
)

+ reasoning_content = response.choices[0].message.reasoning_content
content = response.choices[0].message.content
```

**Comment by arunpatala** [comment]
This would also be great to have when we are trying to get structured outputs from content without trying to parse the reasoning output.

seconded +1

**Comment by gaocegege** [comment]
@arunpatala Hi, could you please explain more about the use case?

**Comment by gaocegege** [comment]
Hey @simon-mo, 

I was thinking it might be cool to create a new abstraction called `Reasoning Parser`, kind of like what we have in [abstract_tool_parser.py](https://github.com/vllm-project/vllm/blob/ce69f7f7542bdb8b6e6302d112fb9fad212c1460/vllm/entrypoints/openai/tool_parsers/abstract_tool_parser.py).

We could have a specific implementation like `DeepSeekR1ReasoningParser` that parses the `<think>` and `</think>` tokens to generate `reasoning_content` for `delta_message` in streaming requests and `message` in sync requests.

And we need to add two CLI arguments to `vllm serve`:

```bash
vllm serve meta-llama/Llama-3.1-8B-Instruct \
    --enable-reasoning \
    --reasoning-parser deepseek_r1 \
```

One key insight is that the `<think>` token usually shows up as the first token in the response. But I'm not sure if that’s consistent across different models in the future, so I’d rather not rely on it for optimization. 

What do you think?

**Comment by gaocegege** [comment]
> from pydantic import BaseModel
> from openai import OpenAI
> 
> 
> class Info(BaseModel):
>     name: str
>     age: int
> 
> 
> client = OpenAI(base_url="http://0.0.0.0:8000/v1", api_key="dummy")
> completion = client.beta.chat.completions.parse(
>     model="meta-llama/Llama-3.1-8B-Instruct",
>     messages=[
>         {"role": "system", "content": "You are a helpful assistant."},
>         {"role": "user", "content": "My name is Cameron, I'm 28. What's my name and age?"},
>     ],
>     response_format=Info
> )

I don’t think it will work because the structured output engine, like grammar, sets the logits for the reasoning tokens to −∞. As a result, the output from the LLMEngine doesn't include the reasoning content.

>  tokens that would violate the required structure are identified as invalid. Their logits are set to −∞, effectively assigning them zero probability after the softmax operation and preserving the relative probabilities of other valid tokens. This ensures that only valid tokens are sampled.

**Comment by gaocegege** [comment]
Here is the proposed changes: https://github.com/vllm-project/vllm/pull/12473

**Comment by lucasalvarezlacasa** [comment]
This is not available yet, isn't it? 
I'm getting the following error when trying to use:

`api_server.py: error: unrecognized arguments: --enable-reasoning --reasoning-parser deepseek_r1`

This is the full command I'm launching:

```
docker run --runtime nvidia --gpus all \
    -v /path/to/weights:/root/.cache/huggingface \
    -p 8000:8000 \
    --ipc=host \
    vllm/vllm-openai:v0.7.0 \
    --model deepseek-ai/DeepSeek-R1-Distill-Llama-8B \
    --trust-remote-code \
    --enable-chunked-prefill \
    --uvicorn-log-level error \
    --gpu-memory-utilization 0.95 \
    --dtype bfloat16 \
    --enable-reasoning --reasoning-parser deepseek_r1 \
    --max-model-len 8192
```

**Comment by gaocegege** [comment]
Hi, it does not work with v0.7.0. Perhaps you could try using `vllm/vllm-openai:latest` instead.

**Comment by lucasalvarezlacasa** [comment]
I tried using latest before "v0.7.0" and it didn't work either. I think this is still not released.

**Comment by DarkLight1337** [comment]
Yeah, it's not released yet. You need to use latest code, not latest release (after 0.7.0)

**Comment by linpan** [comment]
v0.7.2 docker not work

**Comment by gaocegege** [comment]
Please check out the fix https://github.com/vllm-project/vllm/pull/13025.

**Comment by hxt365** [comment]
This is broken in 0.7.3

**Comment by nathan-az** [comment]
Quick follow up to clarify the current state. Am I correct that in `main` right now, it's not possible to using reasoning tags with structured response for the answer (post-reasoning), and that this is unlikely to be supported?

Secondarily - would the easiest way to then handle this to just be with two requests with a stop criteria on `/thinking`, then run a second round of inference using structured outputs? This would be slightly annoying due to not using the chat/messages API, but seems do-able.

**Comment by gaocegege** [comment]
> Quick follow up to clarify the current state. Am I correct that in main right now, it's not possible to using reasoning tags with structured response for the answer (post-reasoning), and that this is unlikely to be supported?

It should work in v0. https://docs.vllm.ai/en/v0.8.2/features/reasoning_outputs.html#structured-output

**Comment by Navanit-nebula** [comment]
how to not allow the model to think as in is there any option to not let the llm think and give direct output. Like is there any extra parameter we have or not ?

**Comment by gaocegege** [comment]
> how to not allow the model to think as in is there any option to not let the llm think and give direct output. Like is there any extra parameter we have or not ?

You can add `\n\n</think>` to the chat template, then the model will not think.

**Comment by Navanit-nebula** [comment]
> > how to not allow the model to think as in is there any option to not let the llm think and give direct output. Like is there any extra parameter we have or not ?
> 
> You can add `\n\n</think>` to the chat template, then the model will not think.

@gaocegege I am  inferencing using vllm online serving’s chat completion endpoint as per documentation

https://docs.vllm.ai/en/stable/features/reasoning_outputs.html

```
%%time
import os
import openai

client = openai.OpenAI(
    base_url="http://localhost:8000/v1/", 
    api_key="dummy"
)

chat_completion = client.chat.completions.create(
    messages = [
                {
                'role': 'system',
                'content': "You are an assistant"
            },
            {
                'role': 'user',
                'content': "Tell me joke on bear"
                
            }

            ],
  
    model= "Qwen/QwQ-32B-AWQ",
    temperature=0.0,
    stream=False,
)
reasoning_content = chat_completion.choices[0].message.reasoning_content
content = chat_completion.choices[0].message.content

print("reasoning_content:", reasoning_content)
print("content:", content)
```

How to do so now ?

**Comment by GarrickLin** [comment]
I'm following the instrunctions from [structured-output](https://docs.vllm.ai/en/stable/features/reasoning_outputs.html#structured-output) and just run the sample code 

```python
from openai import OpenAI
from pydantic import BaseModel

# Modify OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8000/v1"

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

class People(BaseModel):
    name: str
    age: int


json_schema = People.model_json_schema()

prompt = ("Generate a JSON with the name and age of one random person.")
completion = client.chat.completions.create(
    model="qwq-32b",
    messages=[{
        "role": "user",
        "content": prompt,
    }],
    extra_body={"guided_json": json_schema},
)
print("reasoning_content: ", completion.choices[0].message.reasoning_content)
print("content: ", completion.choices[0].message.content)
```

the output is 

```
reasoning_content:  {
  "name": "John Doe",
  "age": 30
}
content:  None
```

Is that as expected? The formatted json is in `reasoning_content` instead of `content`. @gaocegege

**Comment by gaocegege** [comment]
> Is that as expected?

No, I think you should specify v0 engine to run it:

```
VLLM_USE_V1=0 vllm serve deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B \
    --enable-reasoning --reasoning-parser deepseek_r1
``` @GarrickLin

**Comment by cyr1x** [comment]
I ve the same behavior with Qwen3 models and V0 engine, when I turn off reasoning, it fills reasoning_content and never content. same with streaming

**Comment by gaocegege** [comment]
Could you give it a new try with the new upstream main branch?

**Comment by tomasruizt** [comment]
Hi! I'm doing some research on reasoning and would like to "condition" the model on a pre-existing reasoning trace. Do you think there is any way to do this with vLLM in general? I'm thinking of something like:

```python
messages = [
    {
        "role": "user",
        "content": "9.11 and 9.8, which is greater?",
    },
    {
        "role": "assistant",
        "reasoning_content": "Clearly, 11 is larger than 8, so 9.11 is larger than 9.8. The answer is therefore 9.11",
        "content": "",   # ideally the model could start generating here
    },
]

response = client.chat.completions.create(
    model=model, messages=messages, max_tokens=100
)
```

**Comment by gaocegege** [comment]
Maybe it is possible with offline inference. I don't think it works with the OpenAI API.
